superflore/utils.py

```diff
@@ -145,67 +144,525 @@ def trim_string(string, length=80):
+                       "Spencer-94",
+                       "Spencer-99",
+                       "SPL-1.0",
+                       "SSH-OpenSSH",
+                       "SSH-short",
+                       "SSPL-1.0",
+                       "SugarCRM-1.1.3",
+                       "SWL",
+                       "TAPR-OHL-1.0",
+                       "TCL",
+                       "TCP-wrappers",
+                       "TMate",
+                       "TORQUE-1.1",
+                       "TOSL",
+                       "TU-Berlin-1.0",
+                       "TU-Berlin-2.0",
+                       "UCL-1.0",
+                       "Unicode-DFS-2015",
+                       "Unicode-DFS-2016",
+                       "Unicode-TOU",
+                       "Unlicense",
+                       "UPL-1.0",
+                       "Vim",
+                       "VOSTROM",
+                       "VSL-1.0",
+                       "W3C",
+                       "W3C-19980720",
+                       "W3C-20150513",
+                       "Watcom-1.0",
+                       "Wsuipa",
+                       "WTFPL",
+                       "X11",
+                       "Xerox",
+                       "XFree86-1.1",
+                       "xinetd",
+                       "Xnet",
+                       "xpp",
+                       "XSkat",
+                       "YPL-1.0",
+                       "YPL-1.1",
+                       "Zed",
+                       "Zend-2.0",
+                       "Zimbra-1.3",
+                       "Zimbra-1.4",
+                       "Zlib",
+                       "zlib-acknowledgement",
+                       "ZPL-1.1",
+                       "ZPL-2.0",
+                       "ZPL-2.1"]
+
+    def map_license_to_spdx(lic):
+        """
+        Map some commonly used license values to one of valid SPDX Identifiers
+        from https://spdx.org/licenses/
+
+        This is mapping only whatever value is listed in package.xml without
+        any knowledge about the actual license used in the source files - it
+        can map only the clear unambiguous cases (while triggering an warning)
+        The rest needs to be fixed in package.xml, so it will trigger an error
+
+        This is similar to what e.g. Openembedded is doing in:
+        http://git.openembedded.org/openembedded-core/tree/meta/conf/licenses.conf
+        """
+        return {
+            'Apache License Version 2.0': 'Apache-2.0',
+            'Apachi 2': 'Apache-2.0',
+            'Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)': 'Apache-2.0',  # noqa: E501
+            'Apache v2': 'Apache-2.0',
+            'Apache v2.0': 'Apache-2.0',
+            'Apache2.0': 'Apache-2.0',
+            'APACHE2.0': 'Apache-2.0',
+            'Apache2': 'Apache-2.0',
+            'Apache License, Version 2.0': 'Apache-2.0',
+            'Apache 2': 'Apache-2.0',
+            'Apache 2.0': 'Apache-2.0',
```

Comment:
@shr-project My conscern is that this will become an unmaintainable, long list of "aliases" for the SPDX identifiers. I really do like the idea of using SPDX, but many of the community packages, unfortunately, fail to be consistent.

This is why I liked the regular expression idea, since it would get "close enough" to correct.

Reply:
There is corresponding PR for catkin_pkg https://github.com/ros-infrastructure/catkin_pkg/pull/296 which shows the license validation warning to the developers and hopefully they will eventually improve the license values in package.xml directly, so this superflore modification would be less and less useful in future until we could drop it completely.

The issue with "close enough" is that in some more "creative" values it can also guess incorrect value and then the bad value from package.xml is still better than wrongly assumed value from regexp.